The aim of this project is to clean the data and analyze the included used car listings.

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos=pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
print("\nColumn Names:")
print(autos.columns)
autos.head()

# analyzing columns having null values:
print("\n% missing values in vehicleType column: ",((50000-44905)*100)/50000)
print("% missing values in gearbox column: ",((50000-47320)*100)/50000)
print("% missing values in model column: ",((50000-47242)*100)/50000)
print("% missing values in fuelType column: ",((50000-45518)*100)/50000)
print("% missing values in notRepairedDamage column: ",((50000-40171)*100)/50000)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Observations : 
* vehicleType, gearbox, model, fuelType, notRepairedDamage columns have null values that need to be inspected.
* price column, odometer have string values, when they should be in integers for performing operations.
* datetime related columns are in strings, they could be converted to datetime datatypes for performing operations.
* Columns having null values have more than ~20% null values.
* some column values are in German, they might need to be translated to English for better clarity.
* The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Modifying Column Names:

In [5]:
column_rename_dict={
    'yearOfRegistration' : 'registration_year',
    'monthOfRegistration' : 'registration_month',
    'notRepairedDamage' : 'unrepaired_damage',
    'dateCreated' : 'ad_created',
}

def camelcase_to_snakecase(str):
    res = [str[0].lower()]
    for c in str[1:]:
        if c in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'):
            res.append('_')
            res.append(c.lower())
        else:
            res.append(c)
     
    return ''.join(res)

autos.rename(columns=column_rename_dict,inplace=True)
autos.rename(mapper=camelcase_to_snakecase,axis='columns',inplace=True)
#checking to see the changes:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Changes made:
* The column names use camelcase instead of Python's preferred snakecase, so those columns were converted to be in snakecase.
* Some column names were not very accurately describing what they represent, so those columns had their names modified based on the data dictionary provided.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Any columns that have mostly one value include : seller,offer_type

Cleaning price and odometer columns:

In [7]:
#price column:
print(autos["price"].value_counts())
print(autos["price"].unique())

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$1,795        1
$67,911       1
$43,461       1
$15,290       1
$6,039        1
Name: price, Length: 2357, dtype: int64
['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [8]:
autos["price"]=autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int32

In [9]:
#odometer column:
print(autos["odometer"].value_counts())
print(autos["odometer"].unique())

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64
['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [10]:
autos["odometer"]=autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
autos.rename(columns={"odometer":"odometer_km"}, inplace=True)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns. 

In [11]:
#price column:
print(autos["price"].unique().shape)
print(autos["price"].describe())
autos["price"].value_counts().sort_index()

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

Since prices of cars for sale cannot be zero, those entries having their selling price as 0 must be removed:

In [12]:
autos=autos.loc[autos["price"]!=0]
print(autos["price"].value_counts())
print("The number of entries remaining in the dataset is ",autos.shape[0])

500      781
1500     734
2500     643
1000     639
1200     639
        ... 
15199      1
1659       1
23450      1
15470      1
69993      1
Name: price, Length: 2356, dtype: int64
The number of entries remaining in the dataset is  48579


In [13]:
# odometer_km column:
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts().sort_index())

count     48579.000000
mean     125766.174685
std       39792.936148
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
5000        837
10000       253
20000       762
30000       780
40000       816
50000      1014
60000      1155
70000      1217
80000      1415
90000      1734
100000     2116
125000     5058
150000    31422
Name: odometer_km, dtype: int64


Let's now move on to the date columns('date_crawled','ad_created' and 'last_seen') and understand the date range the data covers.

In [14]:
print(autos["date_crawled"].describe())
#distribution of 'date_crawled' in percentage
date_crawled_distribution=autos["date_crawled"].str[:10].value_counts(normalize=True,dropna=False).sort_index()*100
print(date_crawled_distribution)

count                   48579
unique                  46895
top       2016-03-16 21:50:53
freq                        3
Name: date_crawled, dtype: object
2016-03-05    2.531958
2016-03-06    1.403899
2016-03-07    3.600321
2016-03-08    3.332716
2016-03-09    3.310072
2016-03-10    3.217440
2016-03-11    3.256551
2016-03-12    3.692954
2016-03-13    1.566521
2016-03-14    3.653842
2016-03-15    3.427407
2016-03-16    2.960127
2016-03-17    3.163919
2016-03-18    1.290681
2016-03-19    3.476811
2016-03-20    3.787645
2016-03-21    3.740299
2016-03-22    3.301838
2016-03-23    3.221557
2016-03-24    2.933366
2016-03-25    3.159802
2016-03-26    3.219498
2016-03-27    3.108339
2016-03-28    3.485045
2016-03-29    3.412997
2016-03-30    3.367710
2016-03-31    3.184504
2016-04-01    3.367710
2016-04-02    3.546800
2016-04-03    3.859692
2016-04-04    3.651784
2016-04-05    1.309208
2016-04-06    0.317009
2016-04-07    0.139978
Name: date_crawled, dtype: float64


Observations on the 'date_crawled' column :
* most of the car sales were crawled in the month of March in 2016

In [15]:
print(autos["ad_created"].describe())
ad_created_distribution=autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100
print(ad_created_distribution)

count                   48579
unique                     76
top       2016-04-03 00:00:00
freq                     1887
Name: ad_created, dtype: object
2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
                ...   
2016-04-03    3.884394
2016-04-04    3.688837
2016-04-05    1.181581
2016-04-06    0.325243
2016-04-07    0.125569
Name: ad_created, Length: 76, dtype: float64


Observations on the 'ad_created' column :


In [16]:
print(autos["last_seen"].describe())
last_seen_distribution=autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100
print(last_seen_distribution)

count                   48579
unique                  38483
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object
2016-03-05     0.107042
2016-03-06     0.432286
2016-03-07     0.539328
2016-03-08     0.743119
2016-03-09     0.961321
2016-03-10     1.066304
2016-03-11     1.237160
2016-03-12     2.379629
2016-03-13     0.889273
2016-03-14     1.259804
2016-03-15     1.587106
2016-03-16     1.644744
2016-03-17     2.807798
2016-03-18     0.734885
2016-03-19     1.582989
2016-03-20     2.064678
2016-03-21     2.064678
2016-03-22     2.136726
2016-03-23     1.852652
2016-03-24     1.976163
2016-03-25     1.920583
2016-03-26     1.679738
2016-03-27     1.564462
2016-03-28     2.087322
2016-03-29     2.235534
2016-03-30     2.476379
2016-03-31     2.379629
2016-04-01     2.278762
2016-04-02     2.492847
2016-04-03     2.519607
2016-04-04     2.447560
2016-04-05    12.476585
2016-04-06    22.182425
2016-04-07    13.190885
Name: last_seen, dtype: float64


Observations on the 'last_seen' column :
* more than 50% of the car sales ads were last seen in the month of April in 2016

Understanding the distribution of the 'registration_year' column:

In [17]:
print(autos["registration_year"].describe())
print(autos["registration_year"].value_counts(dropna=False).sort_index())

count    48579.000000
mean      2004.753000
std         88.631663
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64
1000    1
1001    1
1111    1
1800    2
1910    5
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64


Observations on the 'registration_year' column :
* the earliest year of registering a vehicle was in 1000, during which cars did not exist
* the most recent year during which a car was registered is the year 9999, which has not occured yet.
* clearly, these and other erronous values in this column need to be made good.
   * Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [18]:
listings_of_cars_first_registered_before_1900_or_after_2016 = autos.loc[(autos["registration_year"]<1900) | (autos["registration_year"]>2016)]
listings_of_cars_first_registered_before_1900_or_after_2016["registration_year"].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
2017    1393
2018     471
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

Listings having possibly correct registration years(between 1900 and 2016, both inclusive) for cars on sale:

In [19]:
autos=autos.loc[autos["registration_year"].between(1900,2016)]
print(autos.shape[0]) # number of listings 
autos.head()

46693


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_p_s   model  odometer_km  registration_month fuel_type  \
0    manuell        158  andere       150000                   3       lpg   
1  automatik        286     7er       150000                   6    benzin   
2    manuell        102    golf        70000                   7    benzin   
3  automatik         71  fortwo        70000                   6    benzin   
4    manuell          0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Calculating the distribution of the remaining listings for the column 'registration_year'(in percentages):

In [20]:
registration_years_of_properly_listed_cars=autos["registration_year"].value_counts(normalize=True).sort_index()*100
print(registration_years_of_properly_listed_cars)

1910    0.010708
1927    0.002142
1929    0.002142
1931    0.002142
1934    0.004283
          ...   
2012    2.805560
2013    1.719744
2014    1.422055
2015    0.839526
2016    2.612811
Name: registration_year, Length: 78, dtype: float64


Observations on the listings having correct registration years for the cars listed:

When working with data on cars, it's natural to explore variations across different car brands. We will use aggregation to understand the brand column.

Analyzing whether the choosing top 20 car brands by sales volumes is good for further analysis or if selecting those car brands that account for over a certain percentage of the total car sales (e.g. > 5%) makes sense for proceeding further.:-

In [21]:
brand_distribution=autos["brand"].value_counts()
print(brand_distribution)

volkswagen        9865
bmw               5138
opel              5023
mercedes_benz     4504
audi              4041
ford              3265
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            655
toyota             593
hyundai            468
sonstige_autos     461
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64


Based on frequencies, it looks like 9 car brands (that have over a 1000 of its cars sold at Ebay), are in contention for further analysis.
Lets look at sale numbers in percentages.

In [22]:
normalized_brand_distribution=autos["brand"].value_counts(normalize=True)*100
print(normalized_brand_distribution)

volkswagen        21.127364
bmw               11.003791
opel              10.757501
mercedes_benz      9.645985
audi               8.654402
ford               6.992483
renault            4.713769
peugeot            2.983317
fiat               2.563553
seat               1.826826
skoda              1.640503
nissan             1.526995
mazda              1.518429
smart              1.415630
citroen            1.402780
toyota             1.269998
hyundai            1.002292
sonstige_autos     0.987300
volvo              0.914484
mini               0.875934
mitsubishi         0.822393
honda              0.783843
kia                0.706744
alfa_romeo         0.663911
porsche            0.612512
suzuki             0.593237
chevrolet          0.569679
chrysler           0.351230
dacia              0.263423
daihatsu           0.250573
jeep               0.227015
subaru             0.214165
land_rover         0.209882
saab               0.164907
jaguar             0.156340
daewoo             0

From this, clearly a 5% sales criteria would serve us better when we perform further analysis on car brands.

In [23]:
five_pct_car_brands=normalized_brand_distribution.loc[normalized_brand_distribution>5.0].index
print(five_pct_car_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


Let us now aggregate on these brands to find out the mean price of car sales of these brands on Ebay

In [24]:
brands_mean_car_price={}
for car_brand in five_pct_car_brands:
    brand_listings=autos.loc[autos["brand"]==car_brand]
    mean_brand_price=brand_listings["price"].mean()
    brands_mean_car_price[car_brand]=mean_brand_price

message_template="An {car_brand} car is sold at an average price of {mean_price} euros."

for brand, avg_price in brands_mean_car_price.items():
    print(message_template.format(car_brand=brand,mean_price=avg_price))

An volkswagen car is sold at an average price of 6729.81956411556 euros.
An bmw car is sold at an average price of 8571.480147917478 euros.
An opel car is sold at an average price of 5432.479195699781 euros.
An mercedes_benz car is sold at an average price of 30829.021092362345 euros.
An audi car is sold at an average price of 9336.687453600594 euros.
An ford car is sold at an average price of 7456.547932618683 euros.


A mercedes_benz car sells for a much higher average price by a mile when compared to the other 5 car brands.
In the overall picture though:
* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:
* it's difficult to compare more than two aggregate series objects if we want to extend to more columns
* we can't compare more than a few rows from each series object
* we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Let us now aggregate on these brands to find out the mean mileage for each of these brands on Ebay

In [25]:
brands_mean_car_mileage={}
for car_brand in five_pct_car_brands:
    brand_listings=autos.loc[autos["brand"]==car_brand]
    mean_brand_mileage=brand_listings["odometer_km"].mean()
    brands_mean_car_mileage[car_brand]=mean_brand_mileage

message_template="An {car_brand} car is has an average mileage of {mean_mlg}."

for brand, avg_mlg in brands_mean_car_mileage.items():
    print(message_template.format(car_brand=brand,mean_mlg=avg_mlg))


An volkswagen car is has an average mileage of 128713.6340598074.
An bmw car is has an average mileage of 132575.9050214091.
An opel car is has an average mileage of 129314.15488751742.
An mercedes_benz car is has an average mileage of 130792.62877442273.
An audi car is has an average mileage of 129157.38678544914.
An ford car is has an average mileage of 124243.49157733537.


We will now combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly.

In [26]:
#making both the dictionaries into series objects:
bmcp_series=pd.Series(brands_mean_car_price)
print(bmcp_series)
bmcm_series=pd.Series(brands_mean_car_mileage)
print(bmcm_series)

volkswagen        6729.819564
bmw               8571.480148
opel              5432.479196
mercedes_benz    30829.021092
audi              9336.687454
ford              7456.547933
dtype: float64
volkswagen       128713.634060
bmw              132575.905021
opel             129314.154888
mercedes_benz    130792.628774
audi             129157.386785
ford             124243.491577
dtype: float64


In [27]:
#making a dataframe containing both series objects:
brands_mean_price_mileage = pd.DataFrame(bmcp_series, columns=['mean_price'])
print(brands_mean_price_mileage)
brands_mean_price_mileage["mean_mileage"]=bmcm_series
print(brands_mean_price_mileage)

                 mean_price
volkswagen      6729.819564
bmw             8571.480148
opel            5432.479196
mercedes_benz  30829.021092
audi            9336.687454
ford            7456.547933
                 mean_price   mean_mileage
volkswagen      6729.819564  128713.634060
bmw             8571.480148  132575.905021
opel            5432.479196  129314.154888
mercedes_benz  30829.021092  130792.628774
audi            9336.687454  129157.386785
ford            7456.547933  124243.491577


Performing further data cleaning tasks:-

In [28]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_p_s   model  odometer_km  registration_month fuel_type  \
0    manuell        158  andere       150000                   3       lpg   
1  automatik        286     7er       150000                   6    benzin   
2    manuell        102    golf        70000                   7    benzin   
3  automatik         71  fortwo        70000                   6    benzin   
4    manuell          0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [29]:
autos.describe(include='all')

date_crawled                 name  seller offer_type  \
count                 46693                46693   46693      46693   
unique                45136                35822       2          1   
top     2016-04-02 15:49:30  Volkswagen_Golf_1.4  privat    Angebot   
freq                      3                   75   46692      46693   
mean                    NaN                  NaN     NaN        NaN   
std                     NaN                  NaN     NaN        NaN   
min                     NaN                  NaN     NaN        NaN   
25%                     NaN                  NaN     NaN        NaN   
50%                     NaN                  NaN     NaN        NaN   
75%                     NaN                  NaN     NaN        NaN   
max                     NaN                  NaN     NaN        NaN   

               price abtest vehicle_type  registration_year  gearbox  \
count   4.669300e+04  46693        43988       46693.000000    44580   
unique           NaN      2            8                NaN        2   
top              NaN   test    limousine                NaN  manuell   
freq             NaN  24068        12603                NaN    34722   
mean    9.859304e+03    NaN          NaN        2002.908166      NaN   
std     4.912548e+05    NaN          NaN           7.190024      NaN   
min     1.000000e+00    NaN          NaN        1910.000000      NaN   
25%     1.250000e+03    NaN          NaN        1999.000000      NaN   
50%     3.100000e+03    NaN          NaN        2003.000000      NaN   
75%     7.500000e+03    NaN          NaN        2008.000000      NaN   
max     1.000000e+08    NaN          NaN        2016.000000      NaN   

           power_p_s  model    odometer_km  registration_month fuel_type  \
count   46693.000000  44495   46693.000000        46693.000000     43372   
unique           NaN    244            NaN                 NaN         7   
top              NaN   golf            NaN                 NaN    benzin   
freq             NaN   3707            NaN                 NaN     28547   
mean      117.909216    NaN  125581.778853            5.826826       NaN   
std       184.932765    NaN   39857.489693            3.670461       NaN   
min         0.000000    NaN    5000.000000            0.000000       NaN   
25%        75.000000    NaN  100000.000000            3.000000       NaN   
50%       109.000000    NaN  150000.000000            6.000000       NaN   
75%       150.000000    NaN  150000.000000            9.000000       NaN   
max     17700.000000    NaN  150000.000000           12.000000       NaN   

             brand unrepaired_damage           ad_created  nr_of_pictures  \
count        46693             38381                46693         46693.0   
unique          40                 2                   74             NaN   
top     volkswagen              nein  2016-04-03 00:00:00             NaN   
freq          9865             33840                 1821             NaN   
mean           NaN               NaN                  NaN             0.0   
std            NaN               NaN                  NaN             0.0   
min            NaN               NaN                  NaN             0.0   
25%            NaN               NaN                  NaN             0.0   
50%            NaN               NaN                  NaN             0.0   
75%            NaN               NaN                  NaN             0.0   
max            NaN               NaN                  NaN             0.0   

         postal_code            last_seen  
count   46693.000000                46693  
unique           NaN                37155  
top              NaN  2016-04-07 06:17:27  
freq             NaN                    8  
mean    51098.049087                  NaN  
std     25757.049258                  NaN  
min      1067.000000                  NaN  
25%     30827.000000                  NaN  
50%     49828.000000                  NaN  
75%     71735.00000

Columns that contain categorical data in German include 'seller', 'offer_type', 'vehicle_type', 'gearbox', 'fuel_type', and 'unrepaired_damage'. The values shall be mapped to English

In [30]:
autos["seller"].value_counts(dropna=False)

privat        46692
gewerblich        1
Name: seller, dtype: int64

In [31]:
seller_values_map={
    "privat" : "private",
    "gewerblich" : "commercial"
}

In [32]:
autos["offer_type"].value_counts(dropna=False)

Angebot    46693
Name: offer_type, dtype: int64

In [33]:
offer_type_values_map={
   "Angebot" : "offer" 
}

In [34]:
autos["vehicle_type"].value_counts(dropna=False)

limousine     12603
kleinwagen    10585
kombi          8932
bus            4032
cabrio         3016
NaN            2705
coupe          2465
suv            1965
andere          390
Name: vehicle_type, dtype: int64

In [35]:
vehicle_type_values_map={
   "limousine" : "limousine",
   "kleinwagen" : "small car",
    "kombi" : "combi",
    "bus" : "bus",
    "cabrio" : "convertible",
    "coupe" : "coupe",
    "suv" : "suv",
    "andere" : "other"
}

In [36]:
autos["gearbox"].value_counts(dropna=False)

manuell      34722
automatik     9858
NaN           2113
Name: gearbox, dtype: int64

In [37]:
gearbox_values_map={
    "manuell" : "manual",
    "automatik" : "automatic"
}

In [38]:
autos["fuel_type"].value_counts(dropna=False)

benzin     28547
diesel     14034
NaN         3321
lpg          649
cng           71
hybrid        37
elektro       19
andere        15
Name: fuel_type, dtype: int64

In [39]:
fuel_type_values_map={
   "benzin" : "petrol",
    "diesel" : "diesel",
    "lpg" : "lpg",
    "cng" : "cng",
    "hybrid" : "hybrid",
    "elektro" : "electro",
    "andere" : "other"
}

In [40]:
autos["unrepaired_damage"].value_counts(dropna=False)

nein    33840
NaN      8312
ja       4541
Name: unrepaired_damage, dtype: int64

In [41]:
unrepaired_damage_values_map={
    "nein" : "No",
    "ja" : "Yes"
}

In [42]:
#mapping german names to english names:
autos["seller"]=autos["seller"].map(seller_values_map)
autos["offer_type"]=autos["offer_type"].map(offer_type_values_map)
autos["vehicle_type"]=autos["vehicle_type"].map(vehicle_type_values_map)
autos["gearbox"]=autos["gearbox"].map(gearbox_values_map)
autos["fuel_type"]=autos["fuel_type"].map(fuel_type_values_map)
autos["unrepaired_damage"]=autos["unrepaired_damage"].map(unrepaired_damage_values_map)
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    seller offer_type  price   abtest vehicle_type  registration_year  \
0  private      offer   5000  control          bus               2004   
1  private      offer   8500  control    limousine               1997   
2  private      offer   8990     test    limousine               2009   
3  private      offer   4350  control    small car               2007   
4  private      offer   1350     test        combi               2003   

     gearbox  power_p_s   model  odometer_km  registration_month fuel_type  \
0     manual        158  andere       150000                   3       lpg   
1  automatic        286     7er       150000                   6    petrol   
2     manual        102    golf        70000                   7    petrol   
3  automatic         71  fortwo        70000                   6    petrol   
4     manual          0   focus       150000                   7    petrol   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot                No  2016-03-26 00:00:00               0   
1         bmw                No  2016-04-04 00:00:00               0   
2  volkswagen                No  2016-03-26 00:00:00               0   
3       smart                No  2016-03-12 00:00:00               0   
4        ford                No  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Converting the dates to be uniform numeric data(date_crawled, ad_created, last_seen columns):

In [43]:
autos["date_crawled"]=autos["date_crawled"].str[:10].str.replace("-","").astype(int)
autos["ad_created"]=autos["ad_created"].str[:10].str.replace("-","").astype(int)
autos["last_seen"]=autos["last_seen"].str[:10].str.replace("-","").astype(int)
autos.head()

date_crawled                                               name   seller  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2      20160326                         Volkswagen_Golf_1.6_United  private   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0      offer   5000  control          bus               2004     manual   
1      offer   8500  control    limousine               1997  automatic   
2      offer   8990     test    limousine               2009     manual   
3      offer   4350  control    small car               2007  automatic   
4      offer   1350     test        combi               2003     manual   

   power_p_s   model  odometer_km  registration_month fuel_type       brand  \
0        158  andere       150000                   3       lpg     peugeot   
1        286     7er       150000                   6    petrol         bmw   
2        102    golf        70000                   7    petrol  volkswagen   
3         71  fortwo        70000                   6    petrol       smart   
4          0   focus       150000                   7    petrol        ford   

  unrepaired_damage  ad_created  nr_of_pictures  postal_code  last_seen  
0                No    20160326               0        79588   20160406  
1                No    20160404               0        71034   20160406  
2                No    20160326               0        35394   20160406  
3                No    20160312               0        33729   20160315  
4                No    20160401               0        39218   20160401

Seeing if there are particular keywords in the name column that we can extract as new columns:

In [44]:
autos["name"].unique()

array(['Peugeot_807_160_NAVTECH_ON_BOARD',
       'BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik',
       'Volkswagen_Golf_1.6_United', ...,
       'Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon',
       'Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+Reifen_neu_!!',
       'Fiat_500_C_1.2_Dualogic_Lounge'], dtype=object)

All important keywords are already extracted.

Finding the most common brand/model combinations:

In [45]:
brand_model_combinations=autos[["brand","model"]].value_counts(normalize=True)*100
print(brand_model_combinations.iloc[0:20])

brand          model      
volkswagen     golf           8.331273
bmw            3er            5.877065
volkswagen     polo           3.616137
opel           corsa          3.577930
volkswagen     passat         3.031801
opel           astra          3.029554
audi           a4             2.766603
mercedes_benz  c_klasse       2.553096
bmw            5er            2.544106
mercedes_benz  e_klasse       2.153051
audi           a3             1.854141
               a6             1.791212
ford           focus          1.714799
               fiesta         1.622654
volkswagen     transporter    1.517024
renault        twingo         1.382178
peugeot        2_reihe        1.348466
smart          fortwo         1.236094
opel           vectra         1.224857
mercedes_benz  a_klasse       1.211372
dtype: float64


Splitting the odometer_km into groups, and using aggregation to see if average prices follows any patterns based on the mileage:

In [46]:
print(autos["odometer_km"].describe())
print(autos["odometer_km"].value_counts().sort_index())

count     46693.000000
mean     125581.778853
std       39857.489693
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
5000        786
10000       241
20000       742
30000       760
40000       798
50000       995
60000      1128
70000      1187
80000      1375
90000      1673
100000     2059
125000     4858
150000    30091
Name: odometer_km, dtype: int64


splitting the values into groups, each group having a width of 10,000 km and using aggregation to see if average prices follows any patterns based on the mileage.

In [47]:
odometer_groups_avg_prices={}
odometer_milestones=list(range(0,160000,10000))
for i in np.arange(0,len(odometer_milestones)-1,1):
    odometer_range_listings=autos.loc[autos["odometer_km"].between(odometer_milestones[i],odometer_milestones[i+1])]
    odometer_range_mean_price=odometer_range_listings["price"].mean()
    odometer_groups_avg_prices[str(odometer_milestones[i])+"-"+str(odometer_milestones[i+1])+"km"]=odometer_range_mean_price
ogap_series=pd.Series(odometer_groups_avg_prices)
print(ogap_series)
ogap_series.describe()

0-10000km          15392.861733
10000-20000km      18963.915565
20000-30000km      17517.633822
30000-40000km      33567.454429
40000-50000km      36699.373675
50000-60000km      18886.897786
60000-70000km      11637.516631
70000-80000km      10280.345043
80000-90000km       9032.042323
90000-100000km     10959.024116
100000-110000km    12985.474017
110000-120000km             NaN
120000-130000km     6418.588720
130000-140000km             NaN
140000-150000km     7977.449403
dtype: float64


count       13.000000
mean     16178.352097
std       9342.637562
min       6418.588720
25%      10280.345043
50%      12985.474017
75%      18886.897786
max      36699.373675
dtype: float64

Determining how much cheaper cars are with damage than their non-damaged counterparts:-

In [48]:
autos["unrepaired_damage"].unique()
damaged_car_listings = autos.loc[autos["unrepaired_damage"] == 'Yes']
avg_damaged_car_price=damaged_car_listings["price"].mean()
non_damaged_car_listings = autos.loc[autos["unrepaired_damage"] == 'No']
avg_non_damaged_car_price=non_damaged_car_listings["price"].mean()
cost_difference_btw_undamaged_and_damaged=avg_non_damaged_car_price-avg_damaged_car_price
print("Cars with their damages not repaired are on an average",cost_difference_btw_undamaged_and_damaged,"euros cheaper than cars having their damages repaired")

Cars with their damages not repaired are on an average 5873.492165445068 euros cheaper than cars having their damages repaired
